### PR TITLE
[Validator] Explained how to use array constant in annotation

### PR DESCRIPTION
**Pull request by nikophil**
Hi,

few days ago i tried to add a new feature to the `choice` validator constraint, by adding a new `choicesFromConstant` option.
https://github.com/symfony/symfony/pull/29658
But then, i realized this was already supported.

I think this should be documented, as i think a lot of people don't know this feature.

thanks.

### CODE REVIEW

File: reference/constraints/Choice.rst [hunk 1]
```diff
@@ -279,6 +279,34 @@ you can pass the class name and the method as an array.
...
+        protected $genre;
+    }
+
+.. warning::
+
+  Notice that the constant in the option is used without quotes
+
+.. note::
+
+  If the constant is stored in a different class, you can pass the fully qualified class name
```
Comment (HeahDude) [inline reference/constraints/Choice.rst]: What about `defined` instead of `stored`?

**Review by HeahDude — COMMENTED** [review]
Thank you very much for this contribution!

**Comment by nikophil** [comment]
Hi @HeahDude 
i think you know what i'm talking about here ;)

i've wondered if this would need a `configuration-block` and if we have to list all the other way to configure this, but i think it's really specific to the annotation...
what do you think ?

File: reference/constraints/Choice.rst [hunk 2]
```diff
@@ -279,6 +279,34 @@ you can pass the class name and the method as an array.
...
         }
 
+Supplying the Choices from an Array Constant
+--------------------------------------------
+
+You can also directly provide an array constant to the ``choices`` option in the annotation::
+
+    // src/AppBundle/Entity/Author.php
+    namespace AppBundle\Entity;
+
```
Comment (OskarStark) [inline reference/constraints/Choice.rst]: you should add `use Symfony\Component\Validator\Constraints as Assert;` here

File: reference/constraints/Choice.rst [hunk 3]
```diff
@@ -279,6 +279,34 @@ you can pass the class name and the method as an array.
...
+    class Author
+    {
+        const GENRES = ['fiction', 'non-fiction'];
```
Comment (OskarStark) [inline reference/constraints/Choice.rst]: I would prefer sth like `action` and `comedy`

File: reference/constraints/Choice.rst [hunk 4]
```diff
@@ -279,6 +279,34 @@ you can pass the class name and the method as an array.
...
+        protected $genre;
+    }
+
+.. warning::
+
+  Notice that the constant in the option is used without quotes
+
+.. note::
+
+  If the constant is stored in a different class, you can pass the fully qualified class name
```
Comment (OskarStark) [inline reference/constraints/Choice.rst]: s/stored/defined/

File: reference/constraints/Choice.rst [hunk 5]
```diff
@@ -279,6 +279,34 @@ you can pass the class name and the method as an array.
...
+  of the class or import this class by adding it to the "use" list.
```
Comment (OskarStark) [inline reference/constraints/Choice.rst]: This line is not needed

**Comment by OskarStark** [comment]
Thank you for your contribution 👍

File: reference/constraints/Choice.rst (hunk 3 above)
Comment (nikophil) [reply reference/constraints/Choice.rst]: yep, i agree, but i've chosen these one for the coherence with the example above
would you mind that i change the other example as well ?

**Review by OskarStark — APPROVED** [review]

File: reference/constraints/Choice.rst [hunk 6]
```diff
@@ -279,6 +279,35 @@ you can pass the class name and the method as an array.
...
+
+        /**
+         * @Assert\Choice(choices=Author::GENRES)
+         */
+        protected $genre;
+    }
+
+.. warning::
+
+  Notice that the constant in the option is used without quotes.
```
Comment (OskarStark) [inline reference/constraints/Choice.rst]:
```suggestion
  The constant in the option is used without quotes.
```

File: reference/constraints/Choice.rst [hunk 7]
```diff
@@ -279,6 +279,35 @@ you can pass the class name and the method as an array.
             }
         }
 
+Supplying the Choices from an Array Constant
```
Comment (noniagriconomie) [inline reference/constraints/Choice.rst]: array constant in lower case is better i think, wdyt?

File: reference/constraints/Choice.rst [hunk 8]
```diff
@@ -279,6 +279,35 @@ you can pass the class name and the method as an array.
...
+
+        /**
+         * @Assert\Choice(choices=Author::GENRES)
+         */
+        protected $genre;
+    }
+
+.. warning::
+
+  The constant in the option is used without quotes.
```
Comment (noniagriconomie) [inline reference/constraints/Choice.rst]: what about both inside note instead of warning? warning can be misunderstood

Comment (nikophil) [reply reference/constraints/Choice.rst]: do you mean in the **same** note or two different notes ?

File: reference/constraints/Choice.rst (hunk 7 above)
Comment (nikophil) [reply reference/constraints/Choice.rst]: yes you're right

Comment (nikophil) [reply reference/constraints/Choice.rst]: actually i've wrote like this, to be consistent with the title above which is `Supplying the Choices with a Callback Function`

File: reference/constraints/Choice.rst (hunk 8 above)
Comment (nikophil) [reply reference/constraints/Choice.rst]: do you mean both inside the **same** note ?

**Comment by xabbuh** [comment]
I wonder if it would not be easier to expand the code example that we have in the "Basic usage" section to also contain choices coming from a constant and adding an inline comment there.

File: reference/constraints/Choice.rst (hunk 8 above)
Comment (noniagriconomie) [reply reference/constraints/Choice.rst]: the same i think

**Comment by OskarStark** [comment]
friendly ping @nikophil 

@xabbuh's comment is a legitimate consideration 🤔

**Review by xabbuh — APPROVED** [review]

**Comment by javiereguiluz** [comment]
Nicolas, thanks for this contribution ... and sorry for the late merge!

**Pull request closed without merging**